src/adapter/src/coord/introspection.rs

```diff
@@ -0,0 +1,508 @@
+use mz_storage_client::controller::IntrospectionType;
+use tokio::sync::{mpsc, oneshot};
+use tracing::{info, Span};
+
+use crate::coord::{
+    Coordinator, IntrospectionSubscribeFinish, IntrospectionSubscribeOptimizeMir,
+    IntrospectionSubscribeStage, IntrospectionSubscribeTimestampOptimizeLir, Message, PlanValidity,
+    StageResult, Staged,
+};
+use crate::optimize::Optimize;
+use crate::session::Session;
+use crate::util::ClientTransmitter;
+use crate::{optimize, AdapterError, ExecuteContext, ExecuteContextExtra, ExecuteResponse};
+
+// State tracked about an active introspection subscribe.
+#[derive(Debug, Clone)]
+pub(super) struct IntrospectionSubscribe {
+    /// The ID of the targeted cluster.
+    cluster_id: ClusterId,
+    /// The ID of the targeted replica.
+    replica_id: ReplicaId,
+    /// The spec from which this subscribe was created.
+    spec: &'static SubscribeSpec,
+}
+
+impl Coordinator {
+    /// Installs introspection subscribes on all existing replicas.
+    ///
+    /// Meant to be invoked during coordinator bootstrapping.
+    pub(super) async fn bootstrap_introspection_subscribes(&mut self) {
+        let mut cluster_replicas = Vec::new();
+        for cluster in self.catalog.clusters() {
+            for replica in cluster.replicas() {
+                cluster_replicas.push((cluster.id, replica.replica_id));
+            }
+        }
+
+        for (cluster_id, replica_id) in cluster_replicas {
+            self.install_introspection_subscribes(cluster_id, replica_id)
+                .await;
+        }
+    }
+
+    /// Installs introspection subscribes on the given replica.
+    pub(super) async fn install_introspection_subscribes(
+        &mut self,
+        cluster_id: ClusterId,
+        replica_id: ReplicaId,
+    ) {
+        let dyncfgs = self.catalog().system_config().dyncfgs();
+        if !ENABLE_INTROSPECTION_SUBSCRIBES.get(dyncfgs) {
+            return;
+        }
+
+        for spec in SUBSCRIBES {
+            self.install_introspection_subscribe(cluster_id, replica_id, spec)
+                .await;
+        }
+    }
+
+    async fn install_introspection_subscribe(
+        &mut self,
+        cluster_id: ClusterId,
+        replica_id: ReplicaId,
+        spec: &'static SubscribeSpec,
+    ) {
+        let id = self.allocate_transient_id();
+        info!(
+            %id,
+            %replica_id,
+            type_ = ?spec.introspection_type,
+            "installing introspection subscribe",
+        );
+
+        // Sequencing is performed asynchronously, and the target replica may be dropped before it
+        // completes. To ensure the subscribe does not leak in this case, we need to already add it
+        // to the coordinator state here, rather than at the end of sequencing.
+        let subscribe = IntrospectionSubscribe {
+            cluster_id,
+            replica_id,
+            spec,
+        };
+        self.introspection_subscribes.insert(id, subscribe);
```

Comment:
Hm, we just minted the `GlobalId` at the beginning of the method, so this assert would be equivalent to one that checks that `allocate_transient_id` produces unique IDs. I'm not against adding it if you think it's valuable, but it might be more confusing than helpful.